src/renderers/opengl/texture.rs

```diff
@@ -31,34 +33,68 @@ pub(crate) unsafe fn upload_texture(
     gl.tex_image_2d(
         glow::TEXTURE_2D,
         0,
-        format as i32,
+        glow::RGBA as i32,
         width as i32,
         height as i32,
         0,
-        format,
+        glow::RGBA,
         glow::UNSIGNED_BYTE,
         data,
     );
     texture
 }
 
+/// Loads a texture from memory and uploads it to the GPU.
+pub(crate) fn load_texture(gl: &glow::Context, tex: &Texture) -> glow::NativeTexture {
+    match tex {
+        Texture::Rgba {
+            width,
+            height,
+            data,
+        } => unsafe { upload_texture(gl, *width, *height, Some(data)) },
+    }
+}
 
-/// Loads a TGA texture from memory and uploads it to the GPU.
-///
-/// # Panics
-///
-/// Panics if it couldn't read the texture.
-pub(crate) fn load_texture(gl: &glow::Context, tex: &[u8]) -> glow::NativeTexture {
-    // TODO: accept a ModelTexture to support any format
-    match image::load_from_memory_with_format(tex, image::ImageFormat::Tga).unwrap() {
-        image::DynamicImage::ImageRgba8(ref image) => {
-            let (width, height) = image.dimensions();
-            unsafe { upload_texture(gl, width, height, glow::RGBA, Some(image)) }
-        }
-        image::DynamicImage::ImageRgb8(ref image) => {
-            let (width, height) = image.dimensions();
-            unsafe { upload_texture(gl, width, height, glow::RGB, Some(image)) }
-        }
-        image => todo!("Unsupported image: {:?}", image),
+#[cfg(feature = "parallel-tex-dec")]
+pub fn decode_textures(textures: &mut Vec<CompressedTexture>) -> mpsc::Receiver<(usize, Texture)> {
+    let mut num_threads = std::thread::available_parallelism().unwrap().get();
+    if num_threads > 1 {
+        num_threads -= 1;
+    }
+    if num_threads > textures.len() {
+        num_threads = textures.len();
+    }
+
+    let (tx2, rx2) = mpsc::channel();
+    let mut pipes = Vec::with_capacity(num_threads);
+    for _ in 0..num_threads {
+        let (tx, rx) = mpsc::channel::<(usize, CompressedTexture)>();
+        let tx2 = tx2.clone();
+        std::thread::Builder::new()
+            .name(String::from("Texture Decoder"))
+            .spawn(move || {
+                while let Ok((i, tex)) = rx.recv() {
+                    let tex = tex.decode();
+                    tx2.send((i, tex)).unwrap();
+                }
+            })
+            .unwrap();
+        pipes.push(tx);
+    }
+
+    for ((i, tex), tx) in textures.drain(..).enumerate().zip(pipes.iter().cycle()) {
+        tx.send((i, tex)).unwrap();
+    }
+
+    rx2
+}
+
+#[cfg(not(feature = "parallel-tex-dec"))]
+pub fn decode_textures(textures: &mut Vec<CompressedTexture>) -> mpsc::Receiver<(usize, Texture)> {
+    let (tx, rx) = mpsc::channel();
+    for (i, tex) in textures.drain(..).enumerate() {
+        let tex = tex.decode();
+        tx.send((i, tex)).unwrap();
     }
+    rx
 }
```

Comment:
This shouldn't be in the opengl side.

Reply:
Bruh you're right now that I think about it